### PR TITLE
[terra-form-radio] Fixes issue of `onChange` not getting triggered on Radio buttons

### PR DESCRIPTION
**Pull request by supreethmr**
### Summary

**What was changed:**
Fixes issue of `onChange` not getting triggered when radio buttons are navigated using arrow keys. Added new event handler `onKeyDown` for radio button to listen to keydown events and to update radio button state with Mouse / pointer events of radio buttons.

**Why it was changed:**
radio buttons gets checked as we navigate using arrow keys. since radio button checked state gets changes as we navigate through them correspondent `onChange` should be triggered for radio button. since keydown event was not handled on radio button `onChange` was not triggered.

### Testing

This code changes as been validated with controlled Radio Buttons group example and with Jest Test to ensure onChange is triggered when we press down arrow.

This change was tested using:

- [ ] WDIO
- [X] Jest
- [ ] Visual testing (please attach a screenshot or recording)
- [ ] Other (please describe below)
- [ ] No tests are needed

Tested with Controlled RadioField Example which shows the selected value of radio on Onchange event.

### Reviews

In addition to engineering reviews, this PR needs:

- [ ] UX review
- [ ] Accessibility review
- [ ] Functional review

### Additional Details

**This PR resolves:**

UXPLATFORM-9329 

---

Thank you for contributing to Terra.
@cerner/terra

### CODE REVIEW

File: packages/terra-form-radio/src/Radio.jsx [hunk 1]
```diff
@@ -52,21 +52,26 @@ const propTypes = {
...
   onFocus: PropTypes.func,
   /**
     * The value of the input element.
     */
   value: PropTypes.string,
+  /**
+    * @private
+    * Function to trigger on key press on the radio button. Private functions created to be used by field component.
+    */
+  onKeyDown: PropTypes.func,
```
Comment (sdadn) [inline packages/terra-form-radio/src/Radio.jsx:74]: We should add this prop to the CHANGELOG under the `Added` section with a summary of its usage.
e.g.
```
* Added
  * Added (name) prop to (usage). 
```

**Comment by supreethmr** [comment]
triggering onChange() event from `keydown` event causes inconsistency on onChange event has shown on below gif 

https://github.com/cerner/terra-core/assets/12869809/fcd1d701-6f21-46d4-b8df-b466e3b4dbc5

as we can see since we are invoking onChange event with onkeydown event object while setting focus back to first item, event.currentTarget.value will always return the value of radio button on which key-down is triggered and not the value of currently focused radio button. 

To prevent this inconsistency we are just setting the focus back to top radio-button without changing its state to support cyclic navigation on Safari Browser. Safari browser does not support cyclic navigation of radio buttons when they are in group unlike chrome and edge I think we can document behavior of not selecting first and last radio button on cyclic navigation as a known limitation.

File: packages/terra-core-docs/src/terra-dev-site/doc/form-radio/example/field/ControlledRadioField.jsx [hunk 2]
```diff
@@ -20,6 +20,7 @@ export default class extends React.Component {
   }
 
   handleOnChange(e) {
+    console.log(e.currentTarget.value);
```
Comment (ShettyAkarsh) [inline packages/terra-core-docs/src/terra-dev-site/doc/form-radio/example/field/ControlledRadioField.jsx:23]: please remove this

Comment (supreethmr) [reply packages/terra-core-docs/src/terra-dev-site/doc/form-radio/example/field/ControlledRadioField.jsx]: removed https://github.com/cerner/terra-core/pull/3868/commits/84b1bdffbf48a97543ede12eebbfedba7a355604

File: packages/terra-form-radio/src/Radio.jsx [hunk 3]
```diff
@@ -147,6 +154,13 @@ const Radio = ({
     'inner-ring',
   ]);
 
+  const wrapOnKeydown = (e) => {
+    if (onKeyDown) {
+      onKeyDown(e);
+      onChange(e);
```
[the remark this answers is not in the document]
Comment (supreethmr) [reply packages/terra-form-radio/src/Radio.jsx]: no longer required as we are handling onChange on field Component

File: packages/terra-core-docs/CHANGELOG.md [hunk 4]
```diff
@@ -12,6 +12,7 @@
 * Changed
   * Updated all paginator examples for `terra-paginator` to use new props.
   * Updated imports from `src` to `lib`.
+  * Updated `terra-form-radio-field` example to display selected value.
```
Comment (sdadn) [inline packages/terra-core-docs/CHANGELOG.md:15]: This needs to be moved to the unreleased section.

Comment (supreethmr) [reply packages/terra-core-docs/CHANGELOG.md]: updated : https://github.com/cerner/terra-core/pull/3868/files#diff-0d87855fd9582c564e6ca6a9654fcfa4099b7e717a0704a990d80316f05bdf59

**Comment by supreethmr** [comment]
@cerner/terra-code-owners  All the comments have addressed. Please review the pull request

**Comment by saket2403** [comment]
Can we add a jest test to check if onChange is gettting triggered when using keyboard something like this https://github.com/cerner/terra-core/blob/3b6cf76ad4f4a53ae61d56a319bed376b6531df8/packages/terra-button-group/tests/jest/ButtonGroup.test.jsx#L65-L66

File: packages/terra-form-radio/src/RadioField.jsx [hunk 5]
```diff
@@ -144,34 +145,51 @@ const RadioField = (props) => {
...
-        radioItems[radioItems.length - 1].checked = true;
-      } else {
-        radioItems[itemIndex - 1].focus();
-        radioItems[itemIndex - 1].checked = true;
+    if (radioGroup) {
+      const radioItems = radioGroup.querySelectorAll('[type=radio]');
+      const itemIndex = Array.from(radioItems).indexOf(event.currentTarget);
+      const onClick = new MouseEvent('click', { bubbles: true, cancelable: false });
+      const firstItemIndex = findFirstFocasableItem(radioItems);
+      const lastItemIndex = findLastFocasableItem(radioItems);
```
Comment (saket2403) [inline packages/terra-form-radio/src/RadioField.jsx:159]: nit:
```suggestion
      const lastItemIndex = findLastFocusableItem(radioItems);
```

**Comment by supreethmr** [comment]
> Can we add a jest test to check if onChange is getting triggered when using keyboard something like this
> 
> https://github.com/cerner/terra-core/blob/3b6cf76ad4f4a53ae61d56a319bed376b6531df8/packages/terra-button-group/tests/jest/ButtonGroup.test.jsx#L65-L66

Tried both simulating and dispatching `keydown` event on last radio button in group to validate cyclic navigation but it looks like since I am using  dispatch event to trigger on click on element not able to mock the dispatchEvent added for radio-button instance within radio-field. Hence I have written Jest to validate the existence of callbacks on safari browser [here](https://github.com/cerner/terra-core/pull/3868/files#diff-88b9af0e072df23ec05bdb868c775c70a9841199311194fc1b24068ab078dc0e)

File: packages/terra-form-radio/src/RadioField.jsx (hunk 5 above)
Comment (supreethmr) [reply packages/terra-form-radio/src/RadioField.jsx]: fixed here : https://github.com/cerner/terra-core/pull/3868/commits/2d8cb8bab0993e49a496373626270f3abd96a0e6

**Review by saket2403 — APPROVED** [review]

**Pull request merged**